SCAutolib/controller.py

```diff
@@ -0,0 +1,419 @@
+import yaml
+from pathlib import Path
+from schema import Schema, Use, Or, And, Optional
+from shutil import rmtree
+from typing import Union
+
+from SCAutolib import (logger, run, LIB_DIR, LIB_BACKUP, LIB_DUMP,
+                       LIB_DUMP_USERS, LIB_DUMP_CAS, LIB_DUMP_CARDS)
+from SCAutolib.exceptions import SCAutolibWrongConfig, SCAutolibException
+from SCAutolib.models import CA, file, user, card, authselect as auth
+from SCAutolib.models.file import File
+from SCAutolib.utils import (OSVersion, _check_selinux, _gen_private_key,
+                             _get_os_version, _install_packages,
+                             _check_packages, dump_to_json)
+
+
+class Controller:
+    authselect: auth.Authselect = auth.Authselect()
+    sssd_conf: file.SSSDConf = file.SSSDConf()
+    lib_conf: dict = None
+    _lib_conf_path: Path = None
+    local_ca: CA.LocalCA = None
+    ipa_ca: CA.IPAServerCA = None
+    users: [user.User] = None
+
+    @property
+    def conf_path(self):
+        return self._lib_conf_path
+
+    def __init__(self, config: Union[Path, str], params: {} = None):
+        """
+        Constructor will parse and check input configuration file. If some
+        required fields in the configuration are missing, CLI parameters
+        would be checked if missing values are there. If not, an exception
+        would be raised. After parsing the configuration file and filling
+        internal values of the Controller object, other related objects (users,
+        cards, CAs, Authselect, etc.) would be initialised, but any real action
+        that would affect the system
+        wouldn't be made.
+
+        :param config: Path to configuration file with metadata for testing.
+        :type config: pathlib.Path or str
+        :param params: Parameters from CLI
+        :type params: dict
+        :return:
+        """
+
+        # Check params
+
+        # Parse config file
+        self._lib_conf_path = config.absolute() if isinstance(config, Path) \
+            else Path(config).absolute()
+
+        with self._lib_conf_path.open("r") as f:
+            self.lib_conf = json.load(f)
+            assert self.lib_conf, "Data are not loaded correctly."
+        self.lib_conf = self._validate_configuration(params)
```

Comment:
I mean, `self.lib_conf` is a dict if I recall correctly, and if you assign to it twice..?

Reply:
aha, I get it. If you are talking about line 55 and line 57, then it works in the following way:
1. JSON file is loaded to the standard python dict with no library-specific validations
2. Coorect loading of the JSON file is check by `assert` statment
3. Validation of the loaded configuration in form of dict is made in `_validate_configuration` method.
    - in this `_validate_configuration` there is a definition of the configurations using Python [schema](https://pypi.org/project/schema/) module
    - in this method the configuration in for of dict is validate on [this](https://github.com/x00Pavel/SCAutolib/blob/97f925c34c1188027d66333665f0897d71e0c577/SCAutolib/controller.py#L372) line
    - `.validate` method returns a dictionary with processed values from the original dict. How the values are processed (retyping, validation of each value, etc.) is defined in the schema for each part of the configuration. Check `_validate_configuration` method for more details.
 4. As a result, on the second assigning to the `self.lib_conf` we have there fully parsed, retyped and validated dictionary

Does it make sense now?

Reply:
I think I understand how this works. It is horribly un-intuitive. What about something like:
```
with ...:
    self.lib_conf = json.load(f)
    ...
self.parse_params(params)
if not self._config_is_valid():
    raise ...
```

One more thing: I am starting to doubt about the like 56. Is `assert` the reasonable way to handle data loading? Wouldn't an exception be better?

Reply:
Well, here I have several considerations:

1. If config is not valid, then the exception would be raised by schema module. What actually makes sense, because we use this module to validate the data, and would expect that while parsing if something wrong, an exception would be raised there
2. Parameters from the CLI shouldn't be validated in a different place then whole config is parsed just because those values are passed to the defaults of the config (if necessary) and, again, parsed by the schema module
3. In the library, we rely on specific types in the parsed config. This is achieved also by scheme module, because during parsing it tries to retype the value to specified type. And again, the exception is raised if the value can't be retyped. So, we need to use return value from `schema.validate` method
4. Regarding `assert`, yes, you are right, an exception makes more sense
5. Maybe you were confused by using same variable in two places: in loading JSON file and then in validating the those data. So, to overcome this confusion I suggest following code:
```python3
with json_file.open() as f:
    tmp_var = json.load(f)
self.lib_conf = self._validate_configuration(tmp_var, params)
```
What do you think about this?

Reply:
I agree with this approach, it is much more readable.